flink-formats/flink-avro-glue-schema-registry/src/test/java/org/apache/flink/formats/avro/glue/schema/registry/GlueSchemaRegistryAvroSchemaCoderTest.java

```diff
@@ -0,0 +1,287 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.avro.glue.schema.registry;
+
+import com.amazonaws.services.schemaregistry.caching.AWSSchemaRegistrySerializerCache;
+import com.amazonaws.services.schemaregistry.common.AWSSchemaRegistryClient;
+import com.amazonaws.services.schemaregistry.common.configs.GlueSchemaRegistryConfiguration;
+import com.amazonaws.services.schemaregistry.exception.AWSSchemaRegistryException;
+import com.amazonaws.services.schemaregistry.serializers.GlueSchemaRegistrySerializationFacade;
+import com.amazonaws.services.schemaregistry.utils.AWSSchemaRegistryConstants;
+import org.apache.avro.Schema;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.EnumSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.services.glue.model.EntityNotFoundException;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Field;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyMap;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doCallRealMethod;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
```

Comment:
Per the "Apache Flink Code Style and Quality Guide" (https://flink.apache.org/contributing/code-style-and-quality-common.html) the use of mockito is not recommended in tests. Check the document, and also the slides linked there for more details.

Reply:
We dont have a mock client or docker image for AWS Schema Registry to work with currently. What alternatives do we have here?

Reply:
I fell foul to this on my first contribution too. 

The guidelines are [here](https://flink.apache.org/contributing/code-style-and-quality-common.html), see "Avoid Mockito".

For for KDS integration we have a [Behaviour factory](https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/testutils/FakeKinesisBehavioursFactory.java) that builds mocks for you with Fake behaviours.

If you are struggling on any particular example let me know and I will help.

Reply:
For example getting rid of the mocking in `testReadSchema_withValidParams_succeeds()` is actually fairly easy:

```diff
--- a/flink-formats/flink-avro-glue-schema-registry/src/test/java/org/apache/flink/formats/avro/glue/schema/registry/GlueSchemaRegistryAvroSchemaCoderTest.java
+++ b/flink-formats/flink-avro-glue-schema-registry/src/test/java/org/apache/flink/formats/avro/glue/schema/registry/GlueSchemaRegistryAvroSchemaCoderTest.java
@@ -21,6 +21,7 @@ package org.apache.flink.formats.avro.glue.schema.registry;
 import com.amazonaws.services.schemaregistry.caching.AWSSchemaRegistrySerializerCache;
 import com.amazonaws.services.schemaregistry.common.AWSSchemaRegistryClient;
 import com.amazonaws.services.schemaregistry.common.configs.GlueSchemaRegistryConfiguration;
+import com.amazonaws.services.schemaregistry.deserializers.AWSDeserializer;
 import com.amazonaws.services.schemaregistry.exception.AWSSchemaRegistryException;
 import com.amazonaws.services.schemaregistry.serializers.GlueSchemaRegistrySerializationFacade;
 import com.amazonaws.services.schemaregistry.utils.AWSSchemaRegistryConstants;
@@ -64,7 +65,8 @@ public class GlueSchemaRegistryAvroSchemaCoderTest {
     @Mock private AWSSchemaRegistryClient mockClient;
     @Mock private AwsCredentialsProvider mockCred;
     @Mock private GlueSchemaRegistryConfiguration mockConfigs;
-    @Mock private GlueSchemaRegistryInputStreamDeserializer mockInputStreamDeserializer;
+    private GlueSchemaRegistryInputStreamDeserializer mockInputStreamDeserializer =
+            new MockGlueSchemaRegistryInputStreamDeserializer();

     private static Schema userSchema;
     private static User userDefinedPojo;
@@ -83,6 +85,18 @@ public class GlueSchemaRegistryAvroSchemaCoderTest {
                 8, 116, 101, 115, 116, 0, 20, 0, 12, 118, 105, 111, 108, 101, 116
             };

+    private static class MockGlueSchemaRegistryInputStreamDeserializer
+            extends GlueSchemaRegistryInputStreamDeserializer {
+        public MockGlueSchemaRegistryInputStreamDeserializer() {
+            super((AWSDeserializer) null);
+        }
+
+        @Override
+        public Schema getSchemaAndDeserializedStream(InputStream in) throws IOException {
+            return userSchema;
+        }
+    }
+
```

Reply:
Fixed